### PR TITLE
Feat/plugin

### PR DESCRIPTION
**Pull request by Tedc0819**

### CODE REVIEW

File: lib/models/App.js [hunk 1]
```diff
@@ -335,6 +336,44 @@ class App {
     return this
   }
 
+  loadPlugins() {
+    if (!this.config) {
```
Comment (jas-chen) [inline lib/models/App.js]: can let `loadConfig` check this.config itself

File: lib/models/App.js [hunk 2]
```diff
@@ -335,6 +336,44 @@ class App {
...
+     * it will override the sl-express one
+     */
+
+    const plugins = AppHelper.importPlugins(
+      dirPath,
+      { required: false, only: this.config.app.plugins }
+    )
+
+    const existingKeys = Object.keys(plugins)
+    const remainingKeys = this.config.app.plugins.filter( (key) => {
```
Comment (jas-chen) [inline lib/models/App.js]:
```suggestion
    const remainingKeys = this.config.app.plugins.filter((key) => {
```

Comment (jas-chen) [reply lib/models/App.js]: i think we need a linter lol

File: test/exampleApp/api/plugins/queueTask/lib/QueueTaskPlugin.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  async connectDependencies(app) {
+
+  }
+
+  async disconnectDependencies(app) {
+
+  }
+
+  async willStartService(app) {
+    return 'abc'
```
Comment (jas-chen) [inline test/exampleApp/api/plugins/queueTask/lib/QueueTaskPlugin.js]: 🤔

File: lib/models/App.js (hunk 1 above)
Comment (Tedc0819) [reply lib/models/App.js]: a bit different. this part of code means to ensure config is ready in these phases. Maybe we can move the this.config to a lazy loading method. How is it :)

File: lib/models/App.js (hunk 2 above)
Comment (Tedc0819) [reply lib/models/App.js]: sure. but before a linter, we need somebody to have time for fixing. a bit low priority at the moment. sad

Comment (Tedc0819) [reply lib/models/App.js]: let me fix this by the way. I tested the code in ben.shopline.com project. let me find out why

Comment (Tedc0819) [reply lib/models/App.js]: err... I used to use () even it works without ()

File: lib/models/App.js [hunk 4]
```diff
@@ -523,6 +562,28 @@ class App {
 
   }
 
+  executePluginPhase(phase) {
```
Comment (Seitk) [inline lib/models/App.js]: i think we should default this method to be async, doesn't sound very helpful to keep both lol

File: lib/models/App.js [hunk 5]
```diff
@@ -536,13 +597,16 @@ class App {
...
-
     this.loadConfig()
 
     this.context.Promise = Promise
     this.context.request = request
 
     this.loadFramework()
+    this.loadPlugins()
+
+    this.executePluginPhase('didLoadFramework')
```
Comment (Seitk) [inline lib/models/App.js]: consider wrap the phasing callback with a function to automatically fire will/did/etc events
like 
callbackable('framework', () => {
  this.loadFramework()
  this.loadPlugins()  
})

**Pull request merged**